core/src/main/scala/org/http4s/internal/parsing/Rfc3986.scala

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s.internal.parsing
+
+import cats.parse.Parser.{char, charIn, string1}
+import cats.parse.{Parser, Parser1}
+import cats.parse.Rfc5234.{digit, hexdig}
+import org.http4s.Uri.Ipv6Address
+
+/** Common rules defined in Rfc3986
+  *
+  * @see [[https://tools.ietf.org/html/rfc3986]]
+  */
+private[http4s] object Rfc3986 {
+
+  val ipv6: Parser1[Ipv6Address] = {
+
+    def toIpv6(lefts: collection.Seq[Short], rights: collection.Seq[Short]): Ipv6Address =
+      (lefts ++ collection.Seq.fill(8 - lefts.size - rights.size)(0.toShort) ++ rights) match {
+        case collection.Seq(a, b, c, d, e, f, g, h) =>
+          Ipv6Address(a, b, c, d, e, f, g, h)
+      }
+
+    val h16: Parser1[Short] = {
+      (hexdig ~ hexdig.? ~ hexdig.? ~ hexdig.?).string.map { (s: String) => java.lang.Integer.parseInt(s, 16).toShort }
```

Comment:
The parboiled parser didn't seem to make any of these hex digits optional, but without doing so, I am unsure how IPs like `684D:1111:222:3333:4444:5555:6:77` which contain fewer than 4 chars in a single grouping would parse correctly.

Reply:
Parboiled did it with the `(1 to 4).times` rule.  This is something that's hard to express in cats-parse right now:
https://github.com/typelevel/cats-parse/issues/97.

```
      def h16: Rule1[Short] = rule {
        capture((1 to 4).times(HexDigit)) ~> { (s: String) => java.lang.Integer.parseInt(s, 16).toShort }
      }
```